### ISSUE TITLE
[REVIEW]: seg1d: A Python package for Automated segmentation of one-dimensional (1D) data

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2404]
**Submitting author:** @cadop (<a href="http://orcid.org/0000-0003-3662-7203">Mathew Schwartz</a>)
**Repository:** <a href="https://github.com/cadop/seg1d" target ="_blank">https://github.com/cadop/seg1d</a>
**Version:** 0.1.0
**Editor:** @trallard
**Reviewer:** @AKuederle, @ejhigson
**Archive:** Pending

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/407bdbb225719814e80027876108665e/status.svg)](https://joss.theoj.org/papers/407bdbb225719814e80027876108665e)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/407bdbb225719814e80027876108665e"><img src="https://joss.theoj.org/papers/407bdbb225719814e80027876108665e/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/407bdbb225719814e80027876108665e/status.svg)](https://joss.theoj.org/papers/407bdbb225719814e80027876108665e)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/cadop/seg1d" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@AKuederle & @ejhigson, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @trallard know.

✨ **Please try and complete your review in the next six weeks** ✨ 

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/cadop/seg1d">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@cadop) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [ ] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [ ] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [ ] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/cadop/seg1d">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [ ] **Contribution and authorship:** Has the submitting author (@cadop) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @AKuederle, @ejhigson it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- None

MISSING DOIs

- None

INVALID DOIs

- https://doi.org/10.1137/1.9781611972757.50 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1145/2448196.2448199 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1145/2822013.2822039 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1109/tnsre.2013.2291907 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.3390/s16010066 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.23919/eusipco.2017.8081163 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1109/tnsre.2019.2907483 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1145/956750.956777 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1016/b978-012088469-8.50069-3 is INVALID because of 'https://doi.org/' prefix
- https://dl.acm.org/doi/10.1145/1081870.1081966 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1109/tnsre.2013.2260561 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1109/hic.2016.7797709 is INVALID because of 'https://doi.org/' prefix
- https://doi.org/10.1371/journal.pone.0211466 is INVALID because of 'https://doi.org/' prefix
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02404/joss.02404/10.21105.joss.02404.pdf)

**Comment by AKuederle** [comment]
Started the review by opening issues in the repo:

https://github.com/cadop/seg1d/tree/master/paper

**Comment by cadop** [comment]
@whedon commands

**Comment by whedon** [comment]
Here are some things you can ask me to do:

```
# List Whedon's capabilities
@whedon commands

# List of editor GitHub usernames
@whedon list editors

# List of reviewers together with programming language preferences and domain expertise
@whedon list reviewers

EDITORIAL TASKS

# Compile the paper
@whedon generate pdf

# Compile the paper from alternative branch
@whedon generate pdf from branch custom-branch-name

# Ask Whedon to check the references for missing DOIs
@whedon check references

# Ask Whedon to check repository statistics for the submitted software
@whedon check repository

```

**Comment by cadop** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1137/1.9781611972757.50 is OK
- 10.1145/2448196.2448199 is OK
- 10.1145/2822013.2822039 is OK
- 10.1109/tnsre.2013.2291907 is OK
- 10.3390/s16010066 is OK
- 10.23919/eusipco.2017.8081163 is OK
- 10.1109/tnsre.2019.2907483 is OK
- 10.1145/956750.956777 is OK
- 10.1016/b978-012088469-8.50069-3 is OK
- 10.1145/1081870.1081966 is OK
- 10.1109/tnsre.2013.2260561 is OK
- 10.1109/hic.2016.7797709 is OK
- 10.1371/journal.pone.0211466 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by cadop** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02404/joss.02404/10.21105.joss.02404.pdf)

**Comment by trallard** [comment]
Hi folks just checking on the review.

@ejhigson do you have any updates on the review here? I know these are complicated times and your contribution and time are appreciated.

**Comment by ejhigson** [comment]
@trallard I plan to take a look at this in the next week!

**Comment by ejhigson** [comment]
@cadop congrations on a nice software package and paper! I have installed it and successfully run the tests and examples, as well as reviewing the documentation and paper. I only have a few minor questions:

1) I see the minimum versions for the dependencies in ``setup.py`` are different to those which are listed in the documentation - should this be updated?

2) I get an error running the code in the README as I think you need to change ``seg1d.segmentData`` to ``seg1d.segment_data`` (as in the documentation).

3) I see that there are three authors on the paper, but that you appear to be the only code contributor to the github module. Please can you confirm the role of the other two authors in the software package, and that they meet the [JOSS authorship guidelines](https://joss.readthedocs.io/en/latest/submitting.html#authorship)?

**Comment by AKuederle** [comment]
To provide a summary of my feedback:

First of all, the package is very well written and documented. It fullfils the purpose, which is very well outlined in the paper, nicely and thanks to the examples it should be easy to get it running on a new dataset.

From a scientific perspective, this package mainly wraps a template matching approach based on cross correlation with a subsequent clustering with a nice interface. While, in particular the template matching part could be considered a very common approach to my (and the authors) knowledge, there is no simple implementation out there. Therefore, a nice implementation with any easy-to-use interface is definitely valuable.

However, I during the review, I had a couple of concerns regarding the overall amount of content in the software package:

1. The name `seg1d` suggests that this is a package that implements a set of possible ways to segment a time series, but it just implements a single method.
2. Second, given that it just implements a single method, from a user perspective, I would burden myself with an additional dependency. My suggestion (independent of this paper and review) to author would be, to try to get a modified version of the implementation integrated into one of the already popular timeseries analysis package. This would ensure that people can find and use the method easily, and would increase the chance that the software is properly maintained for a longer time.

Besides the points above, I would fully recommend the package for publication. It is a nice piece of software with a clear usecase.

**Comment by cadop** [comment]
@ejhigson  Thank you for pointing out the readme error!  The commits below correspond to your items.

1) https://github.com/cadop/seg1d/commit/dc007ac4115de3df8e555278d322b33f1f41fafc https://github.com/cadop/seg1d/commit/efe73589648c21f4b1e1fa4c309ae77cd87e746d

2) https://github.com/cadop/seg1d/commit/d01632e3d497aa098b1a31b09e3d7abf2e65da45

3) https://github.com/cadop/seg1d/issues/1

**Comment by ejhigson** [comment]
Hi @cadop - thank you very much for these changes, everything looks good to me! I am now happy to recommend the paper for publication in JOSS. Congratulations on a nice software package!

**Comment by cadop** [comment]
@ejhigson Thank you very much!

@AKuederle Thank you for the in-depth review and time you spent on considering the package for the community.  Please let me know if we need to continue some of the issue discussions as I have left them for you to close in your perspective.  As for the points you raise, I think its overall a fair assessment.  For some clarity, the package name is similar to `mwarp1d` due to an initial goal of combining them in a sort of ecosystem.  

@trallard  I am not sure you received the notification, but you were tagged in this issue https://github.com/cadop/seg1d/issues/1 and I believe it is just waiting for you to close it. 

Please let me know if there are any more steps/actions needed by me.

**Comment by cadop** [comment]
@trallard  It seems that the issue I mentioned above is the only thing left in the checks and it requires your sign-off. Please let me know if there is something waiting on me otherwise.

**Comment by cadop** [comment]
Hi @kthyng , Since you helped in pre-review, I was wondering if you could help move this along. I assume COVID has been taking up peoples time.

**Comment by trallard** [comment]
Apologies for the delay @cadop  I will jump into the issue and will make sure to speed this review on my end

**Comment by trallard** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1137/1.9781611972757.50 is OK
- 10.1145/2448196.2448199 is OK
- 10.1145/2822013.2822039 is OK
- 10.1109/tnsre.2013.2291907 is OK
- 10.3390/s16010066 is OK
- 10.23919/eusipco.2017.8081163 is OK
- 10.1109/tnsre.2019.2907483 is OK
- 10.1145/956750.956777 is OK
- 10.1016/b978-012088469-8.50069-3 is OK
- 10.1145/1081870.1081966 is OK
- 10.1109/tnsre.2013.2260561 is OK
- 10.1109/hic.2016.7797709 is OK
- 10.1371/journal.pone.0211466 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by trallard** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://raw.githubusercontent.com/openjournals/joss-papers/joss.02404/joss.02404/10.21105.joss.02404.pdf)

**Comment by trallard** [comment]
@cadop I have created a [PR](https://github.com/cadop/seg1d/pull/10) to make some minor edits to the paper for acceptance. 

@AKuederle if you're happy with the authoring resolution can you complete your checklist, please?

**Comment by AKuederle** [comment]
@trallard done :)

**Comment by trallard** [comment]
@AKuederle  and @ejhigson  thank you very much for your contribution as reviewers for JOSS. Your time and contributions are much appreciated 🙌🏼

**Comment by trallard** [comment]
@cadop since both reviewers have completed their duties and not flagged any major issues we are ready to move forward. 

At this point could you:
- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

I can then move forward with accepting the submission.

**Comment by cadop** [comment]
@trallard  Thanks for moving all these things along.

I made a release of the software https://github.com/cadop/seg1d/releases/tag/v0.1.2

and made a zenodo upload: 
https://zenodo.org/record/3979649#.XzLab55KiUk

DOI 10.5281/zenodo.3979649


Is this all correct? 

Thanks

**Comment by cadop** [comment]
@AKuederle  @ejhigson  Thank you for the helpful feedback and improvements!

**Comment by trallard** [comment]
@whedon set 10.5281/zenodo.3979649 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3979649" target="_blank">10.5281/zenodo.3979649</a> is the archive.

**Comment by trallard** [comment]
@whedon set v0.1.2 as version

**Comment by whedon** [comment]
OK. v0.1.2 is the version.

**Comment by trallard** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1137/1.9781611972757.50 is OK
- 10.1145/2448196.2448199 is OK
- 10.1145/2822013.2822039 is OK
- 10.1109/tnsre.2013.2291907 is OK
- 10.3390/s16010066 is OK
- 10.23919/eusipco.2017.8081163 is OK
- 10.1109/tnsre.2019.2907483 is OK
- 10.1145/956750.956777 is OK
- 10.1016/b978-012088469-8.50069-3 is OK
- 10.1145/1081870.1081966 is OK
- 10.1109/tnsre.2013.2260561 is OK
- 10.1109/hic.2016.7797709 is OK
- 10.1371/journal.pone.0211466 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1644

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1644, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by trallard** [comment]
@cadop Thanks all seems in order. I have proceeded and completed my editorial duties. The Editor in Chief will take from here for acceptance. 

Thanks everyone for your patience in these busy times and your contribution is much appreciated.

**Comment by cadop** [comment]
Thank you @trallard
